data/tutorials/language/0lg_08_env.md

```diff
@@ -0,0 +1,557 @@
+
+It is very rare to need to use the functions above. However, it is important to understand that changing a function may not need any code refactoring. A function can be given several equivalent forms and changed into another, either using refactoring or using a higher-order function. Since functions are values, currying and uncurrying are operations on those values.
+
+In practice, curried functions are the default form function should take.
+- Allows partial application
+- Less editing, no parenthesis and comas
+- No pattern matching over a tuple
+
+### Functions With Side-Effects
+
+With respect to its type, a function is expected to process input data from its domain and produce a result data from its codomain.
+
+However, some functions either take input data outside of their declared domain or produce data outside of their codomain. These out-of-signature data are called effects, or side effects. Input and output are the most common forms of effects. Input is out-of-domain data and output out-of-codomain data. However, the result of functions returning random numbers or the current time is influenced by external factors, which is also called an effect. The external factor is out-of-domain input. Similarly, any observable phenomena triggered by the computation of a function is out-of-codomain output.
+
+In practice, what is considered an effect is an engineering choice. In most circumstances, I/O are considered as effects, unless they are ignored. Electromagnetic radiation emitted by the processor when computing a function isn't usually considered a relevant effect, except in some security-sensitive contexts. In the OCaml community, as well as in the wider functional programming community, functions are often said to be either pure or impure. The former does not have side effects, the latter does. It is important to remember this always assumes some sort of context, any computation has effects, and what is considered relevant is a choice.
+
+Since, by definition, effects lie outside function types, a function type can't reflect the effects a function may have. However, it is important to document the intended side effects a function may have. Consider the `Unix.time` function. It returns the number of seconds elapsed since Jan 1, 1970.
+```ocaml
+# Unix.time ;;
+- : unit -> float = <fun>
+```
+
+This function does not need to be passed any data to return its result. It is entirely context-dependent, entirely determined by external factors. Therefore, it could be passed anything, which would simply be discarded. Note it has to be passed something to trigger the computation, which is requesting the information to the operating system, as suggested by the module name. Since that function needs some data to trigger the computation but the data is going to be ignored, it makes sense to provide `()`, the unit value, the inhabitant of the type which only has a single value.
```

Comment:
```suggestion
No data needs to be passed to this function in order to return its result. It is entirely context-dependent, i.e., entirely determined by external factors. Therefore,I f it wasI passed anything, it would simply be discarded. Note that it has to be passed something to trigger the computation, since it requests the information from the operating system, as suggested by the module name. Since that function needs some data to trigger the computation but the data is going to be ignored, it makes sense to provide `()`, the unit value and inhabitant of the type that only has a single value.
```

Does this make it more or less clear?

Reply:
Accepted